Add RNG driver

ArielAbreu opened the pull request:
This PR adds a VirtioRNG driver. Exposes `getRand(cb)` and `getRandLength(length, cb)` in `runtime.driver.rng` (though, i'm pretty sure you'll want to change that, seems like a bad place for it). I had to remove the PCI debug info iterator, because for some reason, having it there prevented the PCI lister function from reading the RNG device (in fact, it hung the system and didn't even load the shell). I know that's a bug, but for now that's the alternative I could find. Now someone just needs to provide an interface for libsodium 😄. I'm working on a custom (no libsodium) `randomBytes` implementation.

kesla commented:
Cool!

kesla commented:
Maybe it would make sense to create a runtime-node-crypto that extends https://github.com/dominictarr/sodium-browserify?

ArielAbreu commented:
It does have a couple of useful functions already done, so, i'll see how to add some more functionality to it.

ArielAbreu commented:
I'm gonna make getRand and getRandLength synchronous, since node's randomBytes is synchronous.

ArielAbreu commented:
Nevermind, I'll make a separate, synchronous randomBytes function.

ArielAbreu commented:
@kesla Do you happen to know how to create a synchronous function in runtimejs? (to provide compatibility with Node's randomBytes function, which is synchronous)

ArielAbreu commented:
Nevermind! Since I only place one buffer, I made the `runtime.driver.rng` functions synchronous by using `getBuffer()`. Plus, I added a Node compatible `randomBytes()` function.

kesla commented on js/driver/virtio/rng.js:
```diff
@@ -0,0 +1,95 @@
+          var tmp = new Buffer(u8.length);
+          for (let k in u8) {
+            tmp.writeUInt8(u8[k], k);
+          };
```
I think that the runtimejs-method should return the u8-array and then a node-compability layer (runtime-node-crypto) could convert it to a Buffer.

I think that we should avoid having any dependency on node/browserify code in runtimejs-core.

kesla replied:
@ArielAbreu If you do want to convert a Uint8Array to a Buffer you can do `Buffer(u8)`

ArielAbreu replied:
For some reason, before `new Buffer(u8)` wasn't giving me the expected result, but it is now, so i'll use (and move it to `runtime-node-crypto`).

ArielAbreu commented:
Ok, i'll fix this.

ArielAbreu commented:
@kesla I <i>could</i> use `sodium-browserify`, <b>if</b>
&nbsp; A) It didn't say it can't find the sodium module,
&nbsp; and B) When I fix that, it didn't say `no secure random number generator found`.

SO, I'll try to find workarounds for this. But for now, I may have to use the libsodium module directly and wrap it myself.

ArielAbreu commented:
OK! I've merged `getRand` and `getRandLength`, so now it's `getRand([length], [callback])` (the parameters can switch places, and they're both optional), and i've added `getHybridRand`, which works exactly the same as `getRand`, EXCEPT when the result of getting the buffer is `null`, in which case it falls back to the Multiply-With-Carry RNG algorithm (with a few tweaks in order to return a number less than 256 but not a negative) and returns a UInt8Array like normal. `getHybridRand` is recommended since it works like /dev/random and /dev/urandom. I've also removed randomBytes and i'm going to put it in runtime-node-crypto.

iefserge commented:
@ArielAbreu hey, sorry for the delay, I'll review this PR in a little while.

iefserge commented on js/core/pci/scan.js:
```diff
@@ -340,6 +340,7 @@ var pciSpace = (function(pciAccessorFactoryArg) {
  */
 var codeNameResolver = (function() {
   var classCodes = [
+    'RNG Device',
```
this doesn't look right (Class codes section in http://wiki.osdev.org/PCI). RNG device is probably should be Unclassified here

ArielAbreu replied:
yeah, fixed this.

iefserge commented on js/core/pci/scan.js:
```diff
@@ -366,7 +367,7 @@ var codeNameResolver = (function() {
     classCodeToName: function(code) {
       if ('undefined' === typeof classCodes[code]) {
-        return classCodes[0];
+        return classCodes[1];
```
this should not be needed

ArielAbreu replied:
fixed this, too.

iefserge commented on js/core/pci/scan.js:
```diff
@@ -802,28 +803,8 @@ pciManager.each(function(pciDevice) {
-    classData.className + ' IRQ: ' + vector + ' PIN: ' + pins[devicePin];
-  debug(info);
-});
-
```
that's weird, let's add this back, something is really wrong

ArielAbreu replied:
It'd list all the PCI devices in debug mode, and then when it got to the listPCIDevices function (no matter whether it was in debug or normal) it would just hang.

iefserge replied:
I'll look into this issue

iefserge replied:
@ArielAbreu how can I reproduce this?

ArielAbreu replied:
Just use the js folder from this PR as the runtimejs dependency for the initrd and use the debug option as true in the package.json. When you start it, it should hang after the pci debug info.

ArielAbreu replied:
Oh! And add the PCI debug info printer back into the file.

iefserge commented on js/driver/virtio/index.js:
```diff
@@ -35,3 +41,4 @@ function testDeviceId(deviceId) {
 }
 
 runtime.pci.addDriver(0x1af4, testDeviceId, driver);
+runtime.pci.addDriver(0x1af4, testDeviceId, driver);
```
we should be able to register virtio driver once

ArielAbreu replied:
It would only read the subsystem value once, had to add it twice in order to read the subsystem id again and get the RNG driver.

iefserge replied:
there's a bug somewhere probably

iefserge replied:
@ArielAbreu this should be fixed

iefserge commented on js/driver/virtio/rng.js:
```diff
@@ -0,0 +1,141 @@
+
+  function fillRequestQueue(length) {
+    if (reqQueue.descriptorTable.descriptorsAvailable) {
+      if (!reqQueue.placeBuffers([new Uint8Array(length || 1)], true)) {
```
1 byte buffers is probably not a good idea, maybe fill the queue with fixed-size buffers (like 128 bytes) and then slice a required number of bytes each time

ArielAbreu replied:
I think you're right, but 256 would probably be better, and store the unused buffers into a varible for future usage.

iefserge commented on js/driver/virtio/rng.js:
```diff
@@ -0,0 +1,141 @@
+    s1 = s2;
+    s2 = tmp;
+    return tmp;
+  }
```
why this instead of Math.random()?

ArielAbreu replied:
Well, `Math.random()` isn't cryptographically secure (not even pseudo-randomly).

iefserge replied:
hm, how do you know that prng function is cryptographically secure?

ArielAbreu replied:
It uses seeds generated from the VirtioRNG driver (which <i>are</i> cryptographically secure).

iefserge replied:
Btw Math.random() on linux uses seed from /dev/urandom.
We would have to seed Math.random() from VirtioRNG as well, because there are no other sources of randomness (c++ could export js function for that)

ArielAbreu replied:
Makes sense.

iefserge replied:
we can't fetch buffer immediately to seed prng, `reqQueue.getBuffer` might return null

iefserge replied:
(while virtio processing it)

ArielAbreu replied:
That was an issue when I first created the driver. But, then for some reason it just went away. It's still a potential issue, though.

iefserge replied:
yes, it's a timing issue, i wouldn't rely on it

iefserge replied:
after a bit of research looks like V8's Math.random() uses PRNG that is not cryptographically secure (https://en.wikipedia.org/wiki/Linear_congruential_generator). What PRNG algorithm `prng()` function implements?

iefserge replied:
> Wait, wouldn't the driver have to be written in C++ in order to seed Math.random()?
No, we could call into C++ from JS. Anyway, sorry that was a bad idea to use Math.random

ArielAbreu replied:
Well it uses [MultiplyWithCarry](https://en.wikipedia.org/wiki/Multiply-with-carry) (with a few modifications, like I said) which is <i>techincally</i> not crypto secure ([see this](http://crypto.stackexchange.com/questions/10359/what-stops-the-multiply-with-carry-random-number-generator-from-being-a-cryptogr)), BUT, since it is seeded by true randomness (at the start and every time the buffer return is not null), it's output is not predictable, making it cryptographically secure.

ArielAbreu replied:
Perhaps it would be better to use [VMPC-R](http://www.vmpcfunction.com/vmpcr.htm) which **is** a CSPRNG. It also looks easy to implement.

iefserge commented on js/driver/virtio/rng.js:
```diff
@@ -0,0 +1,141 @@
+    return tmp;
+  }
+
+  runtime.driver.rng = {
```
we should probably create a separate component (/core/random) that manages random data. VirtioRNG would be one possible backend it could use.

Need to think about interface, maybe something like this
```js
var source = runtime.random.addSource('virtio-rng');
source.push(arraybuffer) // push random bytes buffer into this source
```

iefserge commented on js/driver/virtio/rng.js:
```diff
@@ -0,0 +1,141 @@
+  }
+
+  runtime.driver.rng = {
+    getRand: function(length, cb) {
```
I'd say let's follow Linux design and have two separate sources of randomness like /dev/random and /dev/urandom. Application would be able to choose between them.
For example, function will always accept a callback and wait for real randomness (like /dev/random). Math.random() should probably be able to replace `/dev/urandom`

ArielAbreu replied:
Except that Math.random() does *not* use a CSPRNG.

iefserge replied:
sorry, you're right

ArielAbreu replied:
But, then again, neither does the prng() function. The only CSPRNG algorithm i've been able to find is [VMPC-R](http://www.vmpcfunction.com/vmpcr.htm).

iefserge replied:
found this https://github.com/rubycon/isaac.js/blob/master/isaac.js. We would have to modify it to not use global object and not change string prototype though. We would be able to seed it with any value (by default it uses `Math.random()`). Btw, I think nice place for third party js deps is `/js/deps/` directory.

iefserge replied:
I can take care of this if you'd like.

ArielAbreu replied:
Seems great, since it's already implemented in js (what I found would have to be done manually). I'll check it out, see how to use it.

ArielAbreu commented:
Hmm, odd. I'm getting merge conflicts remotely on github.com, but locally, the Github app shows absolutely no conflicts, and neither does the command line.

iefserge commented:
@ArielAbreu you might need to refetch upstream and rebase your branch

ArielAbreu commented:
I've fixed the merge conflict, and added isaac.js, but I still haven't used it.

iefserge commented:
@ArielAbreu  nice, I'm going to fix `Math.random()` because now it uses the same seed everytime (isaac.js uses `Math.random()` for the initial seed). We can re-seed isaac.js using bytes from virtio-rng after that (when/if available). This way we can ensure `getPseudoRandomBytes` works even if virtio-rng is not available.

ArielAbreu commented:
You could also just reset isaac. It clears the seed and everything.

ArielAbreu commented:
@iefserge Just pushed a commit containing a new interface supporting multiple sources of randomness for `runtime.random`, which is used by the VirtioRNG driver to add itself and set itself as the default. I tested it and both `runtime.random.rand()` and `runtime.random.urand()` work as expected. `rand()` always accepts (and expects) a callback, and `urand()` returns immediately and generates randomness if there is not enough.

ArielAbreu commented:
OK, <i>now</i> everything should work as expected.

ArielAbreu commented:
Correction: <i><b>now</b></i> it should work. I made VirtioRNG use `Math.random()` if u8 is null when VirtioRNG is requested for a seed.

iefserge commented on js/core/random/index.js:
```diff
@@ -0,0 +1,168 @@
+  }
+};
+
+sources['none'].init();
```
I'd move `none` source to separate file and let it use `addSource` API the same way virtio rng driver does

ArielAbreu replied:
I'll move it out.

iefserge commented on js/core/random/index.js:
```diff
@@ -0,0 +1,168 @@
+    def = name;
+    return true;
+  },
+  rand: function(length, cb) {
```
rand/urand should probably output Uint8Array instead of regular array. Just curious, do you think we should make it similar to browser api https://developer.mozilla.org/en-US/docs/Web/API/RandomSource/getRandomValues (it doesn't allocate new buffer but instead writes values into existing uint8array provided as a parameter). Maybe name is `getRandomValues`/`getPseudoRandomValues` as well?

ArielAbreu replied:
For `urand()`, it could possibly write the values into it, if the parameter is given, otherwise, it returns a new buffer. And yes i'll, rename the functions.

iefserge replied:
@ArielAbreu  I got the idea. Maybe not fill the entire queue with fixed size buffers but instead push user provided buffer right into virtio-rng queue? this avoids any extra buffer allocation

ArielAbreu replied:
Sounds good, it could work (sort of) like the PCI drivers.

ArielAbreu replied:
@iefserge Except the buffers would have to be returned, since strict mode doesn't allow functions' arguments to be modified. Straight from Mozilla:
> ... strict mode code doesn't alias properties of arguments objects created within it. In normal code within a function whose first argument is arg, setting arg also sets arguments[0], and vice versa (unless no arguments were provided or arguments[0] is deleted). arguments objects for strict mode functions store the original arguments when the function was invoked. arguments[i] does not track the value of the corresponding named argument, nor does a named argument track the value in the corresponding arguments[i].

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode

iefserge replied:
hm, this doesn't work?
```js
function foo(u8) {
  u8[3] = 1;
  return u8;
}
foo(new Uint8Array(10));

ArielAbreu replied:
Well, *that*, yes. But creating a variable and referencing it, no:
```js
function foo(u8) {
  u8[3] = 1;
  return u8;
}

var bar = new UInt8Array(10);
foo(bar);
// bar is still an empty UInt8Array
```

ArielAbreu replied:
Nevermind, I take all of that back. I just tested it and (both Node and runtime) change the variable that was referenced. Weird, it **shouldn't** work, according to the docs. Whatever! I guess we can take the user's variable.

iefserge replied:
Objects are passed by reference (it doesn't copy the buffer), we should be ok.

iefserge commented on js/driver/virtio/rng.js:
```diff
@@ -0,0 +1,116 @@
+  function fillRequestQueue() {
+    while (reqQueue.descriptorTable.descriptorsAvailable) {
+      if (!reqQueue.placeBuffers([new Uint8Array(256)], true)) {
+
```
break?

ArielAbreu replied:
Whoops.

iefserge commented on js/driver/virtio/rng.js:
```diff
@@ -0,0 +1,116 @@
+          break;
+        }
+        for (let obj of u8) {
+          randobj.queue.push(obj);
```
do you think it's a good idea to push entire u8 buffers into the queue and then slice them? should be more compact in memory i guess
```js
var nextBuffer = queue.pop();
nextBuffer.slice(0, 3); // take some bytes
queue.push(nextBuffer.slice(3)); // put back the rest
```

iefserge commented on js/core/random/index.js:
```diff
@@ -0,0 +1,168 @@
+          cb(arr);
+        }
+      }
+      method.realRand(loopit);
```
this could potentially block the system (infinite loop) in case virtio is very slow with buffer processing, virtio should probably rely on IRQ callbacks to fetch new buffers when they're available

ArielAbreu commented:
@iefserge I've fixed a couple of things. I renamed `rand` to `getRandomValues` and `urand` to `getPseudoRandomValues`. Now `getRandomValues` loads a Uint8Array into the method's queue and calls the method's `fillQueue`, which fills up the Uint8Array with randomness. Also, since now randomness is not stored, `getPseudoRandomValues` asks for a seed and generates all the randomness from isaac.js using the seed. `getPseudoRandomValues` also accepts a Uint8Array variable from the user instead of a length (but a length is still usable) and returns it in case they did `getPseudoRandomValues(new Uint8Array(10))`.

iefserge commented on js/driver/virtio/rng.js:
```diff
@@ -0,0 +1,106 @@
+  function fillRequestQueue() {
+    while (reqQueue.descriptorTable.descriptorsAvailable) {
+      if (!reqQueue.placeBuffers([new Uint8Array(256)], true)) {
+        break;
```
Fixed length buffers make code really complicated, can we remove the fillRequestQueue() logic and only place the buffer provided in `getRandomValue()` argument into `reqQueue`?

ArielAbreu replied:
You're right. It'd make things **way** easier. I always forget to double check my code.

ArielAbreu commented:
I've removed the `fillRequestQueue()` function code to use it directly.

iefserge commented on js/driver/virtio/rng.js:
```diff
@@ -0,0 +1,91 @@
+  var randobj = {
+    queue: [],
+    init: function() {},
+    seed: function() {
```
what's the point of having a separate seed function instead of using the `getRandomValues()` buffer for the seed?

ArielAbreu replied:
Since getRandomValues uses a callback (and getPseudoRandomValues must be synchronous), it's a separate function that blocks.

iefserge replied:
we should be able to seed getPseudoRandomValues once as part of init code and then just rely on RNG, right?

ArielAbreu replied:
Yeah, i'll do it.

iefserge commented on js/driver/virtio/rng.js:
```diff
@@ -0,0 +1,91 @@
+        dev.queueNotify(QUEUE_ID_REQ);
+      }
+
+      var j = 0;
```
you might want to use IRQ callback to fetch buffers (buffers might not be immediately available to fetch) https://github.com/runtimejs/runtime/blob/master/js/driver/virtio/net.js#L150-L151

ArielAbreu commented:
Pushed a commit that makes `fillQueue()` use an IRQ callback and makes `init()` in every source return a one-time seed to use for isaac.js instead of requesting one every time `getPseudoRandomValues()` is called.

iefserge commented on js/driver/virtio/rng.js:
```diff
@@ -0,0 +1,96 @@
+        dev.queueNotify(QUEUE_ID_REQ);
+      }
+
+      irq.on(function() {
```
IRQ callback needs to be added only once, there is no way to remove callback from CPU interrupt

iefserge replied:
`irq` thing is like event emitter for the notifications from the device. Virtio device will trigger this event each time it's done with a buffer (or multiple buffers).

ArielAbreu replied:
Well, there is no way to move it out of there, it needs the callback.

iefserge replied:
you can push callbacks into the queue as well, then fetch buffers and fire callbacks in irq handler

ArielAbreu commented:
New commit made that has a cbqueue for callbacks and calls `irq.on` once.

iefserge commented on js/driver/virtio/rng.js:
```diff
@@ -0,0 +1,97 @@
+        dev.queueNotify(QUEUE_ID_REQ);
+      }
+
+      var u8 = reqQueue.getBuffer();
```
I think this 1 bytes request on init no longer needed (and it doesn't wait for randomness using irq), isaac is able to initialize itself using Math.random. The code in `js/core/random/index.js` could make regular `getRandomValues` call with a callback to re-seed isaac. I'd say we can remove init() from the random source interface as well.

iefserge commented on js/driver/virtio/rng.js:
```diff
@@ -0,0 +1,97 @@
+
+      return u8[0];
+    },
+    fillQueue: function(cb) {
```
can we hide the `queue` array and not expose it to the outside? Instead of the `fillQueue` we can do `getRandomValues(u8, cb)` to add a request.

iefserge commented on js/driver/virtio/vring/index.js:
```diff
@@ -55,6 +55,9 @@ VRing.prototype.fetchBuffers = function(fn) {
       if (null === u8) {
+        if (done) {
+          done();
+        }
```
is this used anywhere?

ArielAbreu replied:
No, I forgot to remove that. I *was* going to use it, but then I found another solution.

iefserge commented on js/core/random/index.js:
```diff
@@ -0,0 +1,118 @@
+    def = name;
+    return true;
+  },
+  getRandomValues: function(length, cb, method) {
```
I'd not expose method to user, system should probably be able to figure out the best method to use.

ArielAbreu replied:
Well, if other methods are added and the user wants to use that as default, they should be able to.

iefserge replied:
There is a danger that user could rely on some method that is not always available.

iefserge commented on js/core/random/index.js:
```diff
@@ -0,0 +1,118 @@
+        return vari;
+      }
+    } else {
+      var arr = [];
```
unnecessary array, could use Uint8Array directly

iefserge commented on js/core/random/index.js:
```diff
@@ -0,0 +1,118 @@
+      cb(arr.array);
+    });
+  },
+  getPseudoRandomValues: function(length, vari) {
```
what's the vari argument?

ArielAbreu replied:
An optional argument that references a Uint8Array variable that can be used instead of the length. I should probably rename it to `u8`, so that it's clearer what it's used for.

iefserge replied:
might be confusing to have both length and u8 as arguments. What's the reason to support two versions of API (u8 and length)?

ArielAbreu replied:
compatability for both Node.js and the Browser.

iefserge replied:
sounds good, let's make it one generic argument like `value` then

iefserge replied:
the same for getRandomValues (+callback) i guess

ArielAbreu commented:
I've fixed the stuff in the comments. Only the `default` method is used, one generic argument (for getRandomValues, too), use U8 directly, remove `done`, no queue for buffers (`fillQueue` is now `fillBuffer`, which accepts a buffer directly), callback queue is not exposed either, and isaac.js inits itself and is seeded with the first value of the buffer from every successful `getRandomValues` call.

iefserge commented:
Thanks, I'm going to add few improvements and will try to resolve pciscan issue, then it should be good to merge.

ArielAbreu commented:
Sounds good! Maybe the `pciscan` issue only happens with the `virtio-rng` device. I haven't tested this theory yet, just a hunch.
